### ISSUE TITLE
Provide annex backend for "salted" checksums

### ISSUE THREAD

**Issue by mih** [datalad/datalad#3357]
When handling sensitive data even the mere presence of a file key can leak information. Example: two datasets contain the same patient records (one file per patient). Getting access to one study dataset, allows one to cross-reference a patient (given an identical record) across other datasets/studies -- even if I never got access to the content of those datasets.

A solution would be to encode a dataset's ID into the file keys determined by the annex backend (essentially salting the hashes with the dataset ID). This would give a dataset-specific hash that is different from any other dataset, even if the file content is exactly identical.

Naturally many interesting use cases become impossible with such a backend (by design), so this should not be the default, but we need this as an option for handling sensitive data.

@jbpoline might be interested in this issue. @joeyh might already have a solution for this problem.

**Comment by joeyh** [comment]
Seems like a good use case for
http://git-annex.branchable.com/todo/external_backends/

-- 
see shy jo

**Comment by yarikoptic** [comment]
Not a complete solution, as the salting/custom backend would be, but in heudiconv for _scans.tsv files which contain patient visit date, we just added a column with random salt in it.  So this way it is not possible to brute force guess content of the file (all other information besides the date is known from file list), and thus dates, from checksum alone.  So, if it is possible to "salt" the file with patient records itself, it could provide a solution as well.

**Comment by jbpoline** [comment]
sounds a reasonable solution !

**Comment by mih** [comment]
For me this issues is become more and more sever and prevents sharing of dataset repositories due to the risk of re-identification due to cross-comparison. A heudiconv-specific solution is not a workable solution for me, because the most sensitive information often tends to be neither in BIDS files, nor anyhow neuroimaging-related.

**Comment by jbpoline** [comment]
@yarikoptic: might be worth trying "URL" backend for some use cases - those URL-based keys would then be "salted" by the urls themselves ;-)

**Comment by joeyh** [comment]
The WORM backend is another one that can contain arbitrary data. But, with these you lose the ability to checksum verify the data. As I understand it, the salted hash approach would let you recalculate the hash (given the salt) so it would be possible to re-hash content in order to verify it.

**Comment by joeyh** [comment]
I do wonder though, if the salts are made available as part of the dataset, wouldn't it still be possible to combine datasets to correlate information?

For example, if I have access to dataset A with salt SA including the content of files. And I have access to dataset B with SB, but not access to the content of files. I see that file F in B has a salted hash of H. I can then hash each file from A, using salt SB, and see if I get the same hash H. If so, I know that is the same content as file F.

A way around this would be to store the salt itself as an annexed file (using a regular unsalted hash). Then having access to the git repo for B would not expose its salt.

**Comment by mih** [comment]
@joeyh This is correct. Given the secret/salt is present, and the content is available re-hashing is possible.

**Comment by joeyh** [comment]
If the salt were stored as an annexed file, it would follow that the external backend's content verification would need to be able to fail inteterminately, in case the salt was not available. https://git-annex.branchable.com/todo/external_backends/ has some discussion about making external backends operations able to fail like this, so it would be good to know if that seems something you'll likely need.

**Comment by mih** [comment]
> A way around this would be to store the salt itself as an annexed file (using a regular unsalted hash). Then having access to the git repo for B would not expose its salt.

This is indeed the plan: have the salt be an annexed file.

> If the salt were stored as an annexed file, it would follow that the external backend's content verification would need to be able to fail inteterminately, in case the salt was not available. https://git-annex.branchable.com/todo/external_backends/ has some discussion about making external backends operations able to fail like this, so it would be good to know if that seems something you'll likely need.

From my POV the main goal is to be able to have git-annex operate on keys that do not leak identity information outside a defined context (typically a single repo, but maybe a collection of repos using a shared salt).

My current state of exploration of this problem suggests that salt and payload content do not need to have different access permissions. If a user has access to a file in A, they can compute any hashes they like. To cross-reference with a file in B, they additionally need SB. If they have access to just SB and A, they implicitly have access to all shared content in AB via A alone. Bottom line: If a user is given access to files in A, they can be given access to SA too -- without making things worse. If that is true, any consumer of A has all the pieces to perform content validation. And yes, content validation is good. TBH I have not yet explored scenarios in which it would *not* be available.

**Comment by yarikoptic** [comment]
FOI: Joey announced custom backends done in [8.20200720.1-73-g049807dbb](https://git.kitenet.net/index.cgi/git-annex.git/commit/?id=049807dbb) . I have not looked/tried yet.

**Comment by mih** [comment]
I read through it. My feeling is that it should work for us. Thx @joeyh

**Comment by mih** [comment]
I have finally implemented a working scaffold for out end of the backend protocol. Pretty much a heavily reduced variant of the `AnnexRemote` implementation.

**Comment by jbpoline** [comment]
:1st_place_medal:

**Comment by mih** [comment]
First observation in the context of this issue: The backend seems to have no inherent linkage to the underlying repository. The backend protocol does not support an equivalent of `GETGITDIR` of the external special remote protocol.

This is not an existential problem, because when triggered by DataLad, the necessary information can be provided via environment variables. However, this would somewhat cripple the usability as a plain git-annex extension. That being said, an env variable requirement is not without precedent, even in git-annex itself. The webdav special remote requires credentials to be provided this way.

Anyways, pointing to the provider of the secret from an external source is a requirement. Maybe we should entirely separate the provisioning of a secret from its usage in the backend. So instead of providing the source of the credential, the backend would require the secret to be provided in the environment directly. A datalad wrapper would do that, and it would also take care of implementing the potential transport of the credential source (likely an annexed-file), and also the definition of a scope (e.g., using the same secret for a hierarchy of nested datasets).

**Comment by mih** [comment]
Update: The backend base class has now been incorporated into the -next extension.

Sidenote: While only tangentially related to this technical document, I want to point out that metadata record composition cannot only rely on annex key anonymity. It must simultaneously consider the actually recorded availability information for a key -- in particular URL-type keys. Once it can be known that two different (salted) keys (from two different datasets) point to the same download URL, cross-referencing file identify is still and immediately possible. This would generally suggest being extra careful when linking (salted) file identity to file availability using information that itself is identifiable. This is less of a problem for a standard git-annex network of repo clones, but certainly is for global (content-based) URLs.
